File: src/main/java/tooko/td/client/TdOptions.java
package tooko.td.client;

import tooko.main.Env;
import tooko.td.TdApi;

public class TdOptions {

    private TdApi.TdlibParameters parameters;
    private boolean useTestDc = false;
    private String databaseDirectory = Env.getPath("data");
    private String filesDirectory = Env.getPath("cache/files");
    private boolean useFileDatabase = true;
    private boolean useChatInfoDatabase = true;
    private boolean useMessageDatabase = true;
    private boolean useSecretChats = true;
    private int apiId = 971882;
    private String apiHash = "1232533dd027dc2ec952ba91fc8e3f27";
    private String systemLanguageCode = "en";
    private String deviceModel = "Tooko";
    private String systemVersion = "/";
    private String applicationVersion = "1.0";
    private boolean enableStorageOptimizer = false;
    private boolean ignoreFileNames = false;

    TdApi.TdlibParameters build() {

        if (parameters == null) {

            parameters = new TdApi.TdlibParameters();
            parameters.useTestDc = this.useTestDc;
            parameters.databaseDirectory = this.databaseDirectory;
            parameters.filesDirectory = this.filesDirectory;
            parameters.useFileDatabase = this.useFileDatabase;
            parameters.useChatInfoDatabase = this.useChatInfoDatabase;
            parameters.useMessageDatabase = this.useMessageDatabase;
            parameters.useSecretChats = useSecretChats;
            parameters.apiId = this.apiId;
            parameters.apiHash = this.apiHash;
            parameters.systemLanguageCode = this.systemLanguageCode;
            parameters.deviceModel = this.deviceModel;
            parameters.systemVersion = this.systemVersion;
            parameters.applicationVersion = this.applicationVersion;
            parameters.enableStorageOptimizer = this.enableStorageOptimizer;
            parameters.ignoreFileNames = this.ignoreFileNames;

        }

        return parameters;

    }

    public TdOptions useTestDc(boolean useTestDc) {

        this.useTestDc = useTestDc;

        return this;

    }

    public TdOptions databaseDirectory(String databaseDirectory) {

        this.databaseDirectory = databaseDirectory;

        return this;

    }

    public TdOptions useFileDatabase(boolean useFileDatabase) {

        this.useFileDatabase = useFileDatabase;

        return this;

    }

    public TdOptions filesDirectory(String filesDirectory) {

        this.filesDirectory = filesDirectory;

        return this;

    }

    public TdOptions useChatInfoDatabase(boolean useChatInfoDatabase) {

        this.useChatInfoDatabase = useChatInfoDatabase;

        return this;

    }

    public TdOptions useMessageDatabase(boolean useMessageDatabase) {

        this.useMessageDatabase = useMessageDatabase;

        return this;

    }

    public TdOptions useSecretChats(boolean useSecretChats) {

        this.useSecretChats = useSecretChats;

        return this;

    }

    public TdOptions apiId(int apiId) {

        this.apiId = apiId;

        return this;

    }

    public TdOptions apiHash(String apiHash) {

        this.apiHash = apiHash;

        return this;

    }

    public TdOptions systemLanguageCode(String systemLanguageCode) {

        this.systemLanguageCode = systemLanguageCode;

        return this;

    }

    public TdOptions deviceModel(String deviceModel) {

        this.deviceModel = deviceModel;

        return this;

    }

    public TdOptions systemVersion(String systemVersion) {

        this.systemVersion = systemVersion;

        return this;

    }

    public TdOptions applicationVersion(String applicationVersion) {

        this.applicationVersion = applicationVersion;

        return this;

    }

    public TdOptions enableStorageOptimizer(boolean enableStorageOptimizer) {

        this.enableStorageOptimizer = enableStorageOptimizer;

        return this;

    }

    public TdOptions ignoreFileNames(boolean ignoreFileNames) {

        this.ignoreFileNames = ignoreFileNames;

        return this;

    }

}

File: src/main/java/tookox/core/client/TdOptions.kt
package tookox.core.client

import tooko.main.Env
import tooko.td.TdApi.TdlibParameters

class TdOptions {

    private var useTestDc = false
    private var databaseDirectory = Env.getPath("data")
    private var filesDirectory = Env.getPath("cache/files")
    private var useFileDatabase = true
    private var useChatInfoDatabase = true
    private var useMessageDatabase = true
    private var useSecretChats = true
    private var apiId = 971882
    private var apiHash = "1232533dd027dc2ec952ba91fc8e3f27"
    private var systemLanguageCode = "en"
    private var deviceModel = "Tooko"
    private var systemVersion = "/"
    private var applicationVersion = "1.0"
    private var enableStorageOptimizer = false
    private var ignoreFileNames = false

    fun build(): TdlibParameters {

        return TdlibParameters().also {

            it.useTestDc = useTestDc
            it.databaseDirectory = databaseDirectory
            it.filesDirectory = filesDirectory
            it.useFileDatabase = useFileDatabase
            it.useChatInfoDatabase = useChatInfoDatabase
            it.useMessageDatabase = useMessageDatabase
            it.useSecretChats = useSecretChats
            it.apiId = apiId
            it.apiHash = apiHash
            it.systemLanguageCode = systemLanguageCode
            it.deviceModel = deviceModel
            it.systemVersion = systemVersion
            it.applicationVersion = applicationVersion
            it.enableStorageOptimizer = enableStorageOptimizer
            it.ignoreFileNames = ignoreFileNames

        }
    }

    fun useTestDc(useTestDc: Boolean): TdOptions {
        this.useTestDc = useTestDc
        return this
    }

    fun databaseDirectory(databaseDirectory: String): TdOptions {
        this.databaseDirectory = databaseDirectory
        return this
    }

    fun useFileDatabase(useFileDatabase: Boolean): TdOptions {
        this.useFileDatabase = useFileDatabase
        return this
    }

    fun filesDirectory(filesDirectory: String): TdOptions {
        this.filesDirectory = filesDirectory
        return this
    }

    fun useChatInfoDatabase(useChatInfoDatabase: Boolean): TdOptions {
        this.useChatInfoDatabase = useChatInfoDatabase
        return this
    }

    fun useMessageDatabase(useMessageDatabase: Boolean): TdOptions {
        this.useMessageDatabase = useMessageDatabase
        return this
    }

    fun useSecretChats(useSecretChats: Boolean): TdOptions {
        this.useSecretChats = useSecretChats
        return this
    }

    fun apiId(apiId: Int): TdOptions {
        this.apiId = apiId
        return this
    }

    fun apiHash(apiHash: String): TdOptions {
        this.apiHash = apiHash
        return this
    }

    fun systemLanguageCode(systemLanguageCode: String): TdOptions {
        this.systemLanguageCode = systemLanguageCode
        return this
    }

    fun deviceModel(deviceModel: String): TdOptions {
        this.deviceModel = deviceModel
        return this
    }

    fun systemVersion(systemVersion: String): TdOptions {
        this.systemVersion = systemVersion
        return this
    }

    fun applicationVersion(applicationVersion: String): TdOptions {
        this.applicationVersion = applicationVersion
        return this
    }

    fun enableStorageOptimizer(enableStorageOptimizer: Boolean): TdOptions {
        this.enableStorageOptimizer = enableStorageOptimizer
        return this
    }

    fun ignoreFileNames(ignoreFileNames: Boolean): TdOptions {
        this.ignoreFileNames = ignoreFileNames
        return this
    }
}